src/targets/gpu/kernels/include/migraphx/kernels/pad.hpp

```diff
@@ -32,31 +32,125 @@
 
 namespace migraphx {
 
+enum class pad_mode_t
+{
+    constant = 0,
+    reflect  = 1,
+    edge     = 2
+};
+
+// Helper function to calculate reflected index for left padding
+// Pattern: dist=1 -> 1, dist=2 -> 2, ..., with bouncing
+template <class T>
+__device__ auto reflect_index_left(T dist, T size) -> T
+{
+    if(size == 1)
+        return 0;
+
+    auto period = 2 * (size - 1);
+    auto pos    = (dist - 1) % period;
+    return (pos < size - 1) ? (pos + 1) : (period - 1 - pos);
+}
+
+// Helper function to calculate reflected index for right padding
+// Pattern: dist=1 -> size-2, dist=2 -> size-3, ..., with bouncing
+template <class T>
+__device__ auto reflect_index_right(T dist, T size) -> T
+{
+    if(size == 1)
+        return 0;
+
+    auto period = 2 * (size - 1);
+    auto pos    = (dist - 1) % period;
+    return (pos < size - 1) ? (size - 2 - pos) : (pos - size + 2);
+}
+
+// Apply edge padding: clamp to valid range
+// Uses multi and offsets to determine left vs right padding (avoids unsigned underflow issues)
+template <class IndexType, class MultiType, class OffsetsType, class BoundsType, class Range>
+__device__ void apply_edge_padding(IndexType& input_idx,
+                                   const MultiType& multi,
+                                   const OffsetsType& offsets,
+                                   const BoundsType& input_bounds,
+                                   Range range_multi)
+{
+    for(auto j : range_multi)
+    {
+        if(multi[j] < offsets[j])
+            input_idx[j] = 0;
+        else if(input_idx[j] >= input_bounds[j])
+            input_idx[j] = input_bounds[j] - 1;
+    }
```

Comment:
It would be nice to write this as `return max(input_idx, input_bounds - 1)`. `array2vec` might help here: `max(array2vec(input_idx), array2vec(input_bounds) - 1)`.